lib/charms/grafana_k8s/v1/grafana_source.py

```diff
@@ -0,0 +1,336 @@
+            rel_id: an optional integer specifying the relation ID
+                for the grafana source service, only required if the
+                :class:`GrafanaSourceConsumer` has been instantiated in
+                `multi` mode.
+
+        """
+        rel = self.framework.model.get_relation(self.name, rel_id)
+        rel_id = rel_id if rel_id is not None else rel.id
+
+        unique_source_name = "juju_{}_{}_{}".format(
+            self.charm.model.name,
+            self.charm.app.name,
+            rel_id
+        )
+        if not source_name or self._source_name_in_use(source_name):
+            logger.warning(
+                "'source-name' not specified' or provided name is already in use. "
+                "Using safe default: {}.".format(unique_source_name)
+            )
+            source_name = unique_source_name
+
+        data = {
+            "address": address,
+            "port": port,
+            "source-type": source_type,
+            "source-name": source_name
+        }
+
+        self._update_sources(data, rel_id)
+
+    def remove_source(self, rel_id=None) -> None:
+        """Removes a source relation.
+
+        Args:
+            rel_id: an optional integer specifying the relation ID
+                for the grafana-k8s source service, only required if the
+                :class:`GrafanaSourceConsumer` has been instantiated in
+                `multi` mode.
+        """
+        rel = self.framework.model.get_relation(self.name, rel_id)
+
+        if rel_id is None:
+            rel_id = rel.id
+
+        source = self._stored.sources.pop(rel_id)
+
+        if source is not None:
+            self._stored.sources_to_delete.add(source["source-name"])
+
+        self._update_sources({}, rel_id)
+
+    def list_sources(self) -> List[dict]:
+        """Returns an array of currently valid sources"""
+        sources = []
+        for source in self._stored.sources.values():
+            sources.append(source)
+
+        return sources
+
+    def _set_sources(self, event: RelationJoinedEvent):
+        """
+        On a relation_joined event, check to see whether or not we already have stored data
+        and update the relation data bag if we know about anything
+        """
+        rel_id = event.relation.id
+        if not self._stored.sources.get(rel_id, {}):
+            return
+
+        logger.debug("Setting Grafana data sources: %s", self._stored.sources[rel_id])
+        event.relation.data[self.charm.app]["sources"] = json.dumps(
+            self._stored.sources[rel_id]
+        )
+
+    def _update_sources(self, data: Dict, rel_id: int) -> None:
+        """Updates the event data bucket with new data"""
+        self._stored.sources[rel_id] = data
+        rel = self.framework.model.get_relation(self.name, rel_id)
+
+        logger.debug("Updating Grafana source data to {}".format(data))
+        rel.data[self.charm.app]["sources"] = json.dumps(data)
```

Comment:
I wonder why this is stored in app rather than unit data bag.
On leadership change, there is a considerable time without a leader so sources may be lost if updated during that time.

Reply:
So it's available to all Grafana instances, which are "shared nothing". In the case of a leadership change when we are "really" HA, the same concern is possible, because it will require a database backend, and only the leader should be updating the database.

Reply:
`is_leader` guard missing